File: 50.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (n==0) return 1;
        if (n==1) return x;
        if (n==-1) return 1/x;
        double t = myPow(x*x,n/2);
        if(n%2){
            return n<0? 1/x*t : x*t;
        }
        return t;
    }
};

File: golang/50/50.go
package main

func myPow(x float64, n int) float64 {
	if n < 0 {
		x = 1 / x
		n = -n
	}
	return fastPow(x, n)
}

func fastPow(x float64, n int) float64 {
	if n == 0 {
		return 1.0
	}
	half := fastPow(x, n/2)
	if n&1 == 0 {
		return half * half
	} else {
		return half * half * x
	}
}
